src/core/ts/background-scripts/background.js/telemetry/Telemetry.spec.ts

```diff
@@ -3,32 +3,163 @@
-    assert.isUndefined(await stringTest.testGetValue(pingName));
-
-    console.log("test passed");
+import {
+  modelLoadTime,
+  translationTime,
+  wordsPerSecond,
+} from "./generated/performance";
+
+const testStrVal = "test";
+const pingName = "custom";
+
+const submitAndAssert = async () => {
+  const eventTestValue = await eventTest.testGetValue(pingName);
+  const counterTestValue = await counterTest.testGetValue(pingName);
+  const stringTestValue = await stringTest.testGetValue(pingName);
+  // const fromLangValue = await fromLang.testGetValue(pingName);
+  // const toLangValue = await toLang.testGetValue(pingName);
+  const mostRecentEventTestValue = eventTestValue.slice().pop();
+
+  console.log({
+    eventTestValue,
+    mostRecentEventTestValue,
+    counterTestValue,
+    stringTestValue,
+    // fromLangValue,
+    // toLangValue,
+  });
+
+  // assert.equal(eventTestValue.length, 1);
+  assert.equal(mostRecentEventTestValue.name, "event_test");
+  // assert.equal(counterTestValue, 1);
+  assert.equal(stringTestValue, testStrVal);
+
+  telemetry.submit();
+  assert.isUndefined(await eventTest.testGetValue(pingName));
+  assert.isUndefined(await counterTest.testGetValue(pingName));
+  assert.isUndefined(await stringTest.testGetValue(pingName));
+  // assert.isUndefined(await fromLang.testGetValue(pingName));
+  // assert.isUndefined(await toLang.testGetValue(pingName));
+};
+
+describe("Telemetry", function() {
+  beforeEach(async function() {
+    // TODO: Reset Glean before running each test
+    // await Glean.testResetGlean("org-mozilla-bergamot-in-browser-tests");
+  });
+
+  it("test metrics collected via Glean.js directly", async function() {
+    console.info(this.test.fullTitle());
+
+    eventTest.record();
+    // eventTest.record({ from_lang: "from", to_lang: "to" })
+    counterTest.add();
+    stringTest.set(testStrVal);
+
+    await submitAndAssert();
+  });
+
+  it("test metrics collected via GleanRecorder", async function() {
+    console.info(this.test.fullTitle());
+
+    await gleanRecorder.transaction(async () => {
+      eventTest.record();
+      // eventTest.record({ from_lang: "from", to_lang: "to" })
+      counterTest.add();
+      stringTest.set(testStrVal);
+    });
+
+    await submitAndAssert();
+  });
+
+  it("expect multiple simultaneous async metrics collection not to work when using Glean.js directly", async function() {
+    console.info(this.test.fullTitle());
+
+    const expected = {};
+    const actual = {};
+
+    // Dispatch 10 metrics collections simultaneously
+    await Promise.all(
+      [...Array(10).keys()].map(index => {
+        return (async () => {
+          expected[index] = {
+            from: `from${index}`,
+            to: `to${index}`,
+            $modelLoadTime: `${index}`,
+            $translationTime: `${index}`,
+            $wordsPerSecond: `${index}`,
+          };
+          await telemetry.onTranslationAttemptConcluded(
```

Comment:
Interesting. 

I left this function synchronous first but I could nevertheless see that the pings interleaved using the browser console, so I added GleanRecorder to be sure. However, I couldn't incorporate that version of the function in a test since the testGetValue methods are async and those have to run before each ping submit, so I made this function async with a beforeSubmit callback to be able to complete the test.

<img width="1501" alt="_Users_motin_Dev_Projects_mozilla_bergamot_bergamot-browser-extension_src_core_ts_background-scripts_background_js_telemetry_Telemetry_ts" src="https://user-images.githubusercontent.com/793037/112633105-980ded00-8e41-11eb-86c8-7b497e2d445e.png">

Reverting it to a sync function now results in seemingly correct pings (as you suggest), but leaves it untestable in the current testing regime, since making use of testGetValue re-introduces its async nature. I guess I'll simply have to finish the implementation of testing-by-intercepting-proxy-server and then GleanRecorder can be removed again.

Reply:
Good, I am glad you checked  and it worked. In regards to testing:

- Indeed, testing-by-intercepting-proxy-server is the current way. We have [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1691033) filed to provide a testing uploader, but that is not high in our priority list for now.
- We also have [an accepted proposal](https://docs.google.com/document/d/18JSwvz2HFDHfVB3MrJh_ON3gnMfTF-nHFxaBzXqPR_I/edit?usp=sharing) to provide a testing API for pings which would help you here. Unfortunately, that is even further down in our priority list.

Reply:
Also, related to this, I include a test in Glean's test suite to verify that this behaviour remains correct. See: https://github.com/mozilla/glean.js/pull/144/commits/d39256f86c0ebb64d07221dcde313772626be846. Given this is an internal Glean feature either way, I'd say you don't need to test specifically for it.